### ISSUE TITLE
Post-harmony: Make sure overlays don't scan all nodes/segments

### ISSUE THREAD

**Issue by aubergine10** [krzychu124/Cities-Skylines-Traffic-Manager-President-Edition#415]
While investigating #411 some general issues were discovered by @krzychu124 with the way overlays work, specifically lots of them are scanning all nodes/segments rather than just those on-screen.

In addition, we should try and find ways to avoid screen updates if the camera isn't moving. Maybe the overlays @kvakvs and @krzychu124  are working on could be used?

Issue marked as paused until Harmony branch merged.

**Comment by kvakvs** [comment]
That hash structure built in the `SpeedLimitsTool.cs::ShowSigns` has to be built once and used to determine visible nodes from the camera frustum. Kind of one time calculation.

**Comment by aubergine10** [comment]
As each overlay needs list of on-screen nodes and/or segments, could we create some sort of manager class where we just feed it a definition of what we need to know and it finds that stuff for current camera view? It would then cache, but any movement of camera would invalidate the cache. When the tool is closed, the cache gets discarded.

**Comment by aubergine10** [comment]
Further to my edited comment above, we could possibly have a slider in mod options that sets the max distance from camera that nodes/segments will be included. This would allow users with lower spec rigs to have some control over overlay visibility distance.

> Aside: We should probably rename "General" tab to "UI" and focus everything on that tab on UI stuff. We can easily free up some space if we remove the 'lock menu / button' items - by changing the main button and menubar to be dragged by right click instead of left click, thus no accidental repositioning during normal usage. I already have some decent code for handling that, but stalled as I didn't know how to hook in to saving the updated positions.

**Comment by kvakvs** [comment]
Use the fact that when TM:PE tools activate, the world becomes immutable, i.e. nothing can be added, moved or destroyed, and we can rebuild those helper structures once.

**Comment by krzychu124** [comment]
Right, but each time camera moves you have to find all visible segments/nodes again, so I am not sure what you want to 'cache' 😕

**Comment by aubergine10** [comment]
@krzychu124 Is that the node/segment grid thing you were talking about in chat?

**Comment by krzychu124** [comment]
I hope so, but we'll see.

**Comment by aubergine10** [comment]
We could probably use both tbh, use the 'network grid' to quickly find what we're looking for, but cache it as most frames the camera won't be moving.

We could also 'dampen' camera movement deltas - so the camera has to move by certain amount (or remain stable for certain time) before cache is invalidated. That way we're not hammering the 'network grid' as often.

I wonder if there is already something in game that maintains a network grid or something similar?

**Comment by aubergine10** [comment]
> Use the fact that when TM:PE tools activate, the world becomes immutable, i.e. nothing can be added, moved or destroyed, and we can rebuild those helper structures once.

@kvakvs Only if game is paused; otherwise disasters can still happen while TM:PE is active = roads can be destroyed as a result.

That being the case we'd still need to invalidate 'visible network' cache should any changes occur on the map.

**Comment by aubergine10** [comment]
In the network grid, it would be good to reserve two (or more?) bits for transport category, eg. road or tracked. That way we can filter to general network type in the same iteration as we filter to grid area.
